### PR TITLE
Add grid offset to grid generator for grid search

### PR DESCRIPTION
**Pull request by parul100495**
### What's in the PR?

- Addition of grid offset in the grid generator. This is required for grid search algorithm when we  want the grid to be shifted by a lambda vector provided by the user.

Tests:
- Added unit test for generation of grid around user defined lambda vector.
- Added grid search with grid offset test case.

### CODE REVIEW

File: fairlearn/reductions/_grid_search/_grid_generator.py [hunk 1]
```diff
@@ -20,11 +20,17 @@
...
     """A generator of a grid of points with a bounded L1 norm."""
 
-    def __init__(self, grid_size, grid_limit, pos_basis, neg_basis, neg_allowed, force_L1_norm):
+    def __init__(self, grid_size, grid_limit, pos_basis, neg_basis, neg_allowed, force_L1_norm,
+                 grid_center=None):
         # grid parameters
         self.dim = len(pos_basis.columns)
         self.neg_allowed = neg_allowed
         self.force_L1_norm = force_L1_norm
+        if grid_center is None:
```
Comment (romanlutz) [inline fairlearn/reductions/_grid_search/_grid_generator.py:29]: Wonderful! Only one remark: can you add a comment about grid_center, explaining what it is, or if you feel very generous you can add a docstring explaining all the input params.
@MiroDudik @riedgar-ms what is force_L1_norm for?

File: test/unit/reductions/grid_search/test_grid_generator.py [hunk 2]
```diff
@@ -34,6 +34,31 @@ def test_grid_generator_demographic_parity(grid_size, grid_limit):
...
+@pytest.mark.parametrize("grid_size", [5, 6, 10, 11, 100, 101, 102])
+@pytest.mark.parametrize("grid_limit", [1.0, 2.0, 4.0, 10.0])
+def test_grid_generator_demographic_parity_with_center(grid_size, grid_limit):
+    disparity_moment = DemographicParity()
+    events = [_ALL]
+
+    _index = pd.MultiIndex.from_product([['+', '-'], events, [0, 1]],
+                                        names=[_SIGN, _EVENT, _GROUP_ID])
+    add_val = 0.02
+    grid_center = pd.Series([0, 0, 0, add_val], index=_index)
```
Comment (romanlutz) [inline test/unit/reductions/grid_search/test_grid_generator.py:46]: Other than the grid_center this looks identical to the test case above. Can we instead parametrize on a few different center values
```python
@pytest.mark.parametrize("grid_center",[
    pd.Series([0, 0, 0, 0],
    pd.Series([0, 0, 0, 1],
    pd.Series([0, 1, 0, 0],
    pd.Series([0, 0, 0, 0.25],
    pd.Series([0, 0.25, 0, 0]
])
```
?

Comment (romanlutz) [reply test/unit/reductions/grid_search/test_grid_generator.py:46]: I suppose the same applies below, although the vector is twice as long and we'd need to consider every other component as non-zero in at least some test.

File: test/unit/reductions/grid_search/test_grid_generator.py [hunk 3]
```diff
@@ -92,11 +117,47 @@ def test_grid_generator_equalized_odds(grid_limit, grid_size):
...
+
+def calculate_grid(grid_limit, grid_size, disparity_moment, events, grid_center=None):
     X, y, A = _quick_data()
 
     disparity_moment.load_data(X, y, sensitive_features=A)
 
     return _GridGenerator(grid_size, grid_limit,
                           disparity_moment.pos_basis, disparity_moment.neg_basis,
-                          disparity_moment.neg_basis_present, False).grid
+                          disparity_moment.neg_basis_present, False, grid_center).grid
```
Comment (romanlutz) [inline test/unit/reductions/grid_search/test_grid_generator.py:163]: Perhaps we can also add a test case where your grid with the grid_center param is plugged into GridSearch? That would go into a different test file, though (one from grid search)

File: test/unit/reductions/grid_search/test_grid_generator.py (hunk 2 above)
Comment (parul100495) [reply test/unit/reductions/grid_search/test_grid_generator.py:46]: Sure, I'll do that.

File: fairlearn/reductions/_grid_search/_grid_generator.py (hunk 1 above)
Comment (MiroDudik) [reply fairlearn/reductions/_grid_search/_grid_generator.py:29]: lets rename `grid_center` -> `grid_origin` (the grid is not necessarily symmetric). `force_L1_norm=True` ensures that all points of the grid have the L1 norm equal to `grid_limit`; if `force_L1_norm=False` then grid consists of points whose L1 norm is less than equal to `grid_limit`.

File: fairlearn/reductions/_grid_search/_grid_generator.py [hunk 4]
```diff
@@ -20,11 +20,17 @@
...
+            self.grid_center = pd.Series(np.zeros(pos_basis.shape[0], dtype=np.float64),
```
Comment (MiroDudik) [inline fairlearn/reductions/_grid_search/_grid_generator.py:30]: I would write `pd.Series(0, index=pos_basis.index)`

File: fairlearn/reductions/_grid_search/_grid_generator.py (hunk 1 above)
Comment (MiroDudik) [reply fairlearn/reductions/_grid_search/_grid_generator.py:29]: [this is of course no longer true once we shift the grid to a different origin :-)]

Comment (MiroDudik) [reply fairlearn/reductions/_grid_search/_grid_generator.py:29]: Actually, for renaming, let's do `grid_center` -> `grid_offset`. Hopefully the final word on the renaming.

File: fairlearn/reductions/_grid_search/_grid_generator.py [hunk 5]
```diff
@@ -20,11 +20,36 @@
...
+        If True, the range is doubled.
+        :type neg_allowed: boolean
+
+        The `force_L1_norm`, if True, ensures that all points of the grid have the L1 norm equal
+        to grid_limit. If False, then grid consists of points whose L1 norm is less than equal to
+        grid_limit.
+        :type force_L1_norm: boolean
+
+        The `grid_offset` shifts the whole grid by that value.
+        :type grid_offset: float
```
Comment (romanlutz) [inline fairlearn/reductions/_grid_search/_grid_generator.py:43]: Isn't it rather a vector than a scalar?

File: test/unit/reductions/grid_search/test_grid_generator.py [hunk 6]
```diff
@@ -34,6 +34,34 @@ def test_grid_generator_demographic_parity(grid_size, grid_limit):
     assert np.isclose(expected_grid.values, grid.values).all()
 
 
+@pytest.mark.parametrize("grid_size", [5, 6])
+@pytest.mark.parametrize("grid_limit", [1.0, 2.0])
+@pytest.mark.parametrize("grid_offset", [[0, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0],
```
Comment (romanlutz) [inline test/unit/reductions/grid_search/test_grid_generator.py:39]: does the case with 0 0 0 0 not cover the entire test case above?

Comment (parul100495) [reply test/unit/reductions/grid_search/test_grid_generator.py:39]: Yes. It does. I'll remove it from here.

File: fairlearn/reductions/_grid_search/_grid_generator.py (hunk 5 above)
Comment (parul100495) [reply fairlearn/reductions/_grid_search/_grid_generator.py:43]: Yes. My bad. I'll correct the documentation.

File: test/unit/reductions/grid_search/test_grid_generator.py (hunk 2 above)
Comment (romanlutz) [reply test/unit/reductions/grid_search/test_grid_generator.py]: In the changes I only see new cases, but no consolidation with the existing ones. Were there any issues or do you just need a little more time? Either way is fine, just trying to see whether you need help unblocking this.

File: test/unit/reductions/grid_search/test_grid_generator.py (hunk 6 above)
Comment (romanlutz) [reply test/unit/reductions/grid_search/test_grid_generator.py]: Oh no I meant remove the test case above and just cover it through this much more general test case! Sorry, should have been more explicit :-)

Comment (parul100495) [reply test/unit/reductions/grid_search/test_grid_generator.py]: I misunderstood this earlier. I have fixed it now. This PR is complete from my side.

**Pull request merged**